Haskell/src/Scrabble/Matrix.hs

```diff
@@ -0,0 +1,59 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+module Scrabble.Matrix
+  ( Matrix(..)
+  , ListMatrix
+  , pattern LM
+  , Vec(..)
+  ) where
+
+import Data.Functor.Compose (Compose(..))
+import Data.Maybe (Maybe)
+import qualified Data.Maybe as Maybe
+import Scrabble.Types (Pos(..))
+
+class Vec m where
+  before :: Int -> m a -> m a
+  after  :: Int -> m a -> m a
+
+-- | A generic matrix type 'm', whose rows are represented by the
+-- associated 'Row'.
+class Matrix m where
+  type Row m :: * -> *
+  elemAt  :: Pos p => (m a) -> p -> Maybe a
+  row     :: m a -> Int -> Maybe (Row m a)
+  col     :: m a -> Int -> Maybe (Row m a)
```

Comment:
Oh yeah, I guess you could have a `type Col` associated as well.  Or a different name for `Row`, maybe.